tech-specs/src/explore/design/ibc.md

```diff
@@ -0,0 +1,408 @@
+                    match check_connection_state(connection_id) {
+                        StateChange::Created => {
+                            // "ConnectionOpenInit"
+                            // Assert that the corresponding client exists
+                        }
+                        StateChange::Updated => {
+                            // Assert that the version is compatible
+
+                            // Verify the proofs to check the counterpart ledger's state is expected
+                            //   - The state can be inferred from the own connection state change
+                            //   - Use `ibc-rs::ics03_connection::handler::verify::verify_proofs()`
+                        }
+                        _ => return Err(Error::InvalidStateChange("Invalid state change happened")),
+                    }
+                }
+
+                // channel handshake or closing
+                "channelEnds" => {
+                    // Assert that the port is owend
+
+                    // Assert that the corresponding connection exists
+
+                    // Assert that the version is compatible
+
+                    // Check the channel state change
+                    //   - none => INIT, none => TRYOPEN, INIT => OPEN, TRYOPEN => OPEN, or OPEN => CLOSED
+                    match check_channel_state(channel_id) {
+                        StateChange::Created => {
+                            // "ChanOpenInit"
+                            continue;
+                        }
+                        StateChange::Closed => {
+                            // OPEN => CLOSED
+                            match get_proofs(data_from_tx) {
+                                Some(proofs) => {
+                                    // "ChanCloseConfirm"
+                                    // Verify the proofs to check the counterpart ledger's channel has been closed
+                                    //   - Use `ibc-rs::ics04_connection::handler::verify::verify_channel_proofs()`
+                                }
+                                None => {
+                                    // "ChanCloseInit"
+                                    continue;
+                                }
+                            }
+                        }
+                        StateChange::Updated => {
+                            // Verify the proof to check the counterpart ledger's state is expected
+                            //   - The state can be inferred from the own channel state change
+                            //   - Use `ibc-rs::ics04_connection::handler::verify::verify_channel_proofs()`
+                        }
+                        _ => return Err(Error::InvalidStateChange("Invalid state change happened")),
+                    }
+                }
+
+                "nextSequenceSend" => {
+                    // "SendPacket"
+                    let packet = get_packet(key, data_from_tx)?;
+
```

Comment:
is there a function in `ibc-rs` we can call here?

Reply:
`ChannelReader`'s functions are called here.